### PR TITLE
[📈] NT-512 Added creator dashboard button to project page

### PR DESCRIPTION
**Pull request by eoji**
# 📲 What
Creators can now access the project dashboard for a single project.

# 🤔 Why
Cuz they can't back their own project but they _can_ view their dashboard.

# 🛠 How
- Created `ProjectDashboardData` data class. It contains the current project, the stats envelope, and whether or not the dashboard should be configured for a single project.

## CreatorDashboardHeaderHolder
- Changed `projectAndStats` input in `CreatorDashboardHeaderHolderViewModel` to `configureWith ` that takes a `ProjectDashboardData` object.
- Changed `projectAndStats` output in `CreatorDashboardHeaderHolderViewModel` to `projectDashboardData` output that returns a `ProjectDashboardData` object.
- Added `viewProjectButtonIsGone ` output to `CreatorDashboardHeaderHolderViewModel`. It emits a boolean that returns `true` when the view project button should be visible.
- Added `projectViewHolderDashboardClicked` to `Delegate`

## CreatorDashboardActivity 
- Changed `projectAndStats` output in `CreatorDashboardViewModel` to `projectDashboardData` output that emits a `ProjectDashboardData` object.
- Added `viewProjectButtonIsGone ` output to `CreatorDashboardViewModel`. It emits a boolean that emits `true` when the view project button should be visible.
- `CreateDashboardAdapter` now takes in a `ProjectDashboardData ` instead of a `Pair<Project, ProjectStatsEnvelope>`

## ProjectActivity
- Added `projectDashboardButtonText ` output to `ProjectHolderViewModel`. It emits a string resource `Integer` that returns `View progress` when the project is live and `View dashboard` when the project is not live.
- Added `projectDashboardContainerIsGone ` output to `ProjectHolderViewModel`. It emits a `boolean` that returns `true` when the current user is not the current project's creator.
- Implemented `ProjectViewHolder.Delgate.projectViewHolderDashboardClicked` that calls new `creatorDashboardButtonClicked ` input.

## XML
- `dashboard_funding_view`
  - Removed hardcoded button heights.
  - Decreased top margin of funding section.
- `dashboard_referrer_breakdown_layout`
  - Decreased top margin of referrer breakdown section.
- `dashboard_referrer_stats_view`
  - Decreased top margin of top pledge sources section.
- `dashboard_reward_stats_view`
  - Decreased top margin of top rewards section.
- Created reusable `project_creator_dashboard_header` container for portrait and landscape project layout.

## Tests
I love tests

# 👀 See
| Live project 🐛 | Ended project 🦋 |
| --- | --- |
| screenshot-2019-11-06_150827 | screenshot-2019-11-06_150735 |
| Single project before 🐛 | Single project after 🦋 |
|  | screenshot-2019-11-06_150629 |
| Multiple projects before 🐛 | Multiple project after🦋 |
| screenshot-2019-11-06_150642 | screenshot-2019-11-06_150624 |

# 📋 QA
View a project you've created.

# Story 📖
[NT-512]

[NT-512]: https://dripsprint.atlassian.net/browse/NT-512

### CODE REVIEW

File: app/src/main/java/com/kickstarter/ui/activities/CreatorDashboardActivity.java [hunk 1]
```diff
@@ -145,16 +143,17 @@ private void setProjectsForDropdown(final @NonNull List<Project> projects) {
     this.bottomSheetAdapter.takeProjects(projects);
   }
 
-  private void newProjectClicked() {
-    hideBottomSheet();
+  private void setProjectNameTextViews(final @NonNull String projectName) {
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/ui/activities/CreatorDashboardActivity.java]: alpha

File: app/src/main/java/com/kickstarter/ui/activities/CreatorDashboardActivity.java [hunk 2]
```diff
@@ -145,16 +143,17 @@ private void setProjectsForDropdown(final @NonNull List<Project> projects) {
...
+    this.collapsedToolbarTitle.setText(projectName);
   }
 
   private void hideBottomSheet() {
     this.bottomSheetBehavior.setState(BottomSheetBehavior.STATE_COLLAPSED);
     this.bottomSheetScrim.setVisibility(View.GONE);
   }
 
-  public void toggleBottomSheetClick() {
+  public void showBottomSheet() {
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/ui/activities/CreatorDashboardActivity.java]: alpha

File: app/src/main/java/com/kickstarter/ui/viewholders/ProjectViewHolder.java [hunk 3]
```diff
@@ -80,6 +81,8 @@
   protected @Bind(R.id.play_button_overlay) ImageButton playButton;
   protected @Bind(R.id.pledged) TextView pledgedTextView;
   protected @Bind(R.id.project_action_buttons) @Nullable ViewGroup projectActionButtonsContainer;
+  protected @Bind(R.id.project_dashboard_container) ViewGroup projectDashboardContainer;
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/ui/viewholders/ProjectViewHolder.java]: alpha

File: app/src/main/java/com/kickstarter/ui/viewholders/ProjectViewHolder.java [hunk 4]
```diff
@@ -510,6 +524,11 @@ public void creatorNameClick() {
     this.delegate.projectViewHolderCreatorClicked(this);
   }
 
+  @OnClick(R.id.project_dashboard_button)
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/ui/viewholders/ProjectViewHolder.java]: let's change all the click methods to `onClick`

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardHeaderHolderViewModel.java [hunk 5]
```diff
@@ -62,15 +61,17 @@
...
-    /** time remaining for latest project (no units) */
-    Observable<String> timeRemainingText();
-
     /** Emits when we should start the {@link com.kickstarter.ui.activities.MessageThreadsActivity}. */
     Observable<Pair<Project, RefTag>> startMessageThreadsActivity();
 
     /** Emits when we should start the {@link com.kickstarter.ui.activities.ProjectActivity}. */
     Observable<Pair<Project, RefTag>> startProjectActivity();
+
+    /** time remaining for latest project (no units) */
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardHeaderHolderViewModel.java]: let's properly write this documentation

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardViewModel.java [hunk 6]
```diff
@@ -80,24 +102,31 @@ public ViewModel(final @NonNull Environment environment) {
...
           .filter(p -> p.id() != project.id())
           .toList())
-        .flatMap(listObservable -> listObservable);
+        .flatMap(listObservable -> listObservable)
+        .compose(bindToLifecycle())
+        .subscribe(this.projectsForBottomSheet);
 
-      currentProject
-        .compose(combineLatestPair(projectStatsEnvelope))
+      Observable.combineLatest(currentProject, projectStatsEnvelope, intentHasProjectExtra, ProjectDashboardData::new)
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardViewModel.java]: let's rename to `isViewingSingleProject`

File: app/src/main/java/com/kickstarter/ui/adapters/data/ProjectDashboardData.kt [hunk 7]
```diff
@@ -0,0 +1,6 @@
+package com.kickstarter.ui.adapters.data
+
+import com.kickstarter.models.Project
+import com.kickstarter.services.apiresponses.ProjectStatsEnvelope
+
+data class ProjectDashboardData(val project: Project, val projectStatsEnvelope: ProjectStatsEnvelope, val singleProjectView: Boolean)
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/ui/adapters/data/ProjectDashboardData.kt]: let's rename to `isViewingSingleProject`

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardViewModel.java [hunk 8]
```diff
@@ -112,38 +141,48 @@ public ViewModel(final @NonNull Environment environment) {
         .subscribe(this.koala::trackViewedProjectDashboard);
     }
 
+    private final PublishSubject<Void> backClicked = PublishSubject.create();
+    private final PublishSubject<Void> scrimClicked = PublishSubject.create();
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardViewModel.java]: alpha

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardViewModel.java [hunk 9]
```diff
@@ -112,38 +141,48 @@ public ViewModel(final @NonNull Environment environment) {
...
     private final PublishSubject<Void> projectsListButtonClicked = PublishSubject.create();
 
-    private final BehaviorSubject<Void> openBottomSheet = BehaviorSubject.create();
+    private final BehaviorSubject<Boolean> bottomSheetShouldExpand = BehaviorSubject.create();
     private final BehaviorSubject<Boolean> progressBarIsVisible = BehaviorSubject.create();
-    private final BehaviorSubject<Pair<Project, ProjectStatsEnvelope>> projectAndStats = BehaviorSubject.create();
-    private final Observable<List<Project>> projectsForBottomSheet;
+    private final BehaviorSubject<ProjectDashboardData> projectDashboardData = BehaviorSubject.create();
+    private final BehaviorSubject<List<Project>> projectsForBottomSheet = BehaviorSubject.create();
+    private final BehaviorSubject<String> projectName = BehaviorSubject.create();
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardViewModel.java]: alpha

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardViewModel.java [hunk 10]
```diff
@@ -112,38 +141,48 @@ public ViewModel(final @NonNull Environment environment) {
...
+
+    @Override public void backClicked() {
+      this.backClicked.onNext(null);
     }
-    @Override
-    public void projectsListButtonClicked() {
+    @Override public void projectsListButtonClicked() {
       this.projectsListButtonClicked.onNext(null);
     }
+    @Override public void projectSelectionInput(final @NonNull Project project) {
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardViewModel.java]: alpha

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardViewModel.java [hunk 11]
```diff
@@ -112,38 +141,48 @@ public ViewModel(final @NonNull Environment environment) {
...
-    @Override public @NonNull Observable<Pair<Project, ProjectStatsEnvelope>> projectAndStats() {
-      return this.projectAndStats;
+    @Override public @NonNull Observable<ProjectDashboardData> projectDashboardData() {
+      return this.projectDashboardData;
     }
     @Override public @NonNull Observable<List<Project>> projectsForBottomSheet() {
       return this.projectsForBottomSheet;
     }
-
+    @Override public @NonNull Observable<String> projectName() {
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardViewModel.java]: alpha

File: app/src/main/java/com/kickstarter/viewmodels/ProjectHolderViewModel.java [hunk 12]
```diff
@@ -262,6 +268,21 @@ public ViewModel(final @NonNull Environment environment) {
         .map(enabledAndOverride -> Build.isExternal() ? enabledAndOverride.first : enabledAndOverride.second)
         .distinctUntilChanged();
 
+      final Observable<Boolean> currentUserIsCreator = this.project
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/ProjectHolderViewModel.java]: `userIsCreatorOfProject`

File: app/src/main/java/com/kickstarter/viewmodels/ProjectViewModel.kt [hunk 13]
```diff
@@ -796,6 +809,10 @@ interface ProjectViewModel {
             this.contactCreatorClicked.onNext(null)
         }
 
+        override fun creatorDashboardButtonClicked() {
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/ProjectViewModel.kt]: let's rename all of these inputs to match the delegate methods

File: app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardHeaderHolderViewModelTest.java [hunk 14]
```diff
@@ -37,6 +38,7 @@
   private final TestSubscriber<String> timeRemainingText = new TestSubscriber<>();
```
Comment (eoji) [inline app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardHeaderHolderViewModelTest.java]: alpha

File: app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardHeaderHolderViewModelTest.java [hunk 15]
```diff
@@ -50,51 +52,79 @@ protected void setUpEnvironment(final @NonNull Environment environment) {
...
+    this.vm.inputs.configureWith(new ProjectDashboardData(project, projectStatsEnvelope, false));
 
     // Messages button is gone if current user is not the project creator (e.g. a collaborator).
     this.messagesButtonIsGone.assertValue(true);
   }
 
   @Test
-  public void testOtherProjectsButtonIsGone_isTrue_WhenCollaboratorHas1Project() {
-    final User collaboratorWith1Project = UserFactory.collaborator().toBuilder().memberProjectsCount(1).build();
+  public void testMessagesButtonIsGone_whenCurrentUserIsCreator() {
```
Comment (eoji) [inline app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardHeaderHolderViewModelTest.java]: `isProjectCreator`

File: app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardHeaderHolderViewModelTest.java [hunk 16]
```diff
@@ -211,14 +239,30 @@ public void testTimeRemainingText() {
...
+
+  @Test
+  public void testViewProjectButtonIsGone_whenViewingAllProjects() {
+    setUpEnvironment(environment());
+
+    this.vm.inputs.configureWith(new ProjectDashboardData(ProjectFactory.project(), ProjectStatsEnvelopeFactory.projectStatsEnvelope(), false));
+    this.viewProjectButtonIsGone.assertValue(false);
+  }
+
+  private ProjectDashboardData getDashboardDataForState(final @Project.State String state) {
```
Comment (eoji) [inline app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardHeaderHolderViewModelTest.java]: `forProjectState`

File: app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardViewModelTest.java [hunk 17]
```diff
@@ -28,18 +31,60 @@
...
+  public void testBottomSheetShouldExpand_whenScrimClicked() {
+    setUpEnvironment(environment());
+    this.vm.intent(new Intent());
+    this.vm.inputs.scrimClicked();
+
+    this.bottomSheetShouldExpand.assertValue(false);
+    this.koalaTest.assertNoValues();
+  }
+
+  public void testProjectAndStats_whenIntentDoesNotHaveProjectExtra() {
```
Comment (eoji) [inline app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardViewModelTest.java]: `testProjectDashboardData_whenViewingAllProjects`

File: app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardViewModelTest.java [hunk 18]
```diff
@@ -54,8 +99,24 @@ public void testProjectAndStats() {
...
 
     setUpEnvironment(environment().toBuilder().apiClient(apiClient).build());
-    final Pair<Project, ProjectStatsEnvelope> outputPair = Pair.create(ListUtils.first(projects), projectStatsEnvelope);
-    this.projectAndStats.assertValue(outputPair);
+    this.vm.intent(new Intent());
+    this.projectDashboardData.assertValue(new ProjectDashboardData(Objects.requireNonNull(ListUtils.first(projects)), projectStatsEnvelope, false));
+    this.koalaTest.assertValues(KoalaEvent.VIEWED_PROJECT_DASHBOARD);
+  }
+
+  public void testProjectAndStats_whenIntentHasProjectExtra() {
```
Comment (eoji) [inline app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardViewModelTest.java]: `testProjectDashboardData_whenViewingSingleProject`

File: app/src/main/java/com/kickstarter/viewmodels/ProjectViewModel.kt (hunk 13 above)
Comment (eoji) [reply app/src/main/java/com/kickstarter/viewmodels/ProjectViewModel.kt]: So, I looked into it 🔍 it's moving into refactor territory and Out of Scope™ for this PR

Comment (Scollaco) [reply app/src/main/java/com/kickstarter/viewmodels/ProjectViewModel.kt]: Oh yeah... let's save it for a future Investment Day™️

**Pull request merged**